docs/transaction.md

```diff
@@ -0,0 +1,222 @@
+
+Once the transaction is scheduled, the coordinator starts sending the execution messages. We break each command to one or more micro-ops and each operation corresponds to a single message hop.
+
+For example, "MSET" corresponds to a single micro-op "mset" that has the same semantics, but runs in parallel on all the involved shards.
+
+However, "RENAME" requires two micro-ops: fetching the data from two keys, and then the second hop - deleting/writing a key (depending whether the key is a source or a destination).
+
+When a coordinator schedules an execution, it also specifies whether this execution is the last hop for that command. This is necessary, so that shards could do the  clean-ups when running the last execution request.
+
+Once a coordinator sends the micro-op request to all the shards, it waits for the answer. Only when all shards executed the micro-op and return the result, the coordinator is unblocked and it can proceed to the next hop. The coordinator is allowed to process the intermediary responses from the previous hops in order to define the next execution request.
+
+The shards always execute transactions at the head of the tx-queue. When the last execution hop for that transaction is executed the transaction is removed from the queue and the next one can be executed. This way we maintain the ordering guarantees specified by the scheduling order of the transactions and we maintain the atomicity of operations across multiple shards.
+
+## Multi-op transactions (Redis transactions)
+
+Redis transactions (MULTI/EXEC sequences) are modelled as `N` commands within a DF transaction. In order to avoid ambiguity with terms, we call a Redis transaction -  multi-transaction in Dragonfly.
+
+Dragonfly transactional framework allows running any command as standalone transaction or within a multi-transaction. The complexity of handling both cases is pushed into the transactional framework and a command implementation is not aware of a transaction context. Take for example, redis transaction consisting of 2 commands:
+
+```
+MULTI
+SET x foo
+SET y bar
+EXEC
+```
+
+`SET` command can be implemented with two hops (below is a pseudo-code):
+
+```cpp
+trans->Schedule();   // Hop1
+
+trans->Exec(trans, [](key, val) {
+  dict[key] = val;
+});   // Hop2
+```
+
+With multi-transactions we would run the following code that includes both commands.
+
+```cpp
+// set x foo
+trans->Schedule();
+
+trans->Exec([](key, val) {
+  dict[key] = val;    // x <- foo
+});
+
+// set y bar
+trans->Schedule();
+trans->Exec([](key, val) {
+  dict[key] = val;    // y <- bar
+});
+```
+
+However we want to schedule our mult-transaction only once and then set the arguments for each command accordingly. The transactional framework tracks the progress of multi-transaction - it exits early from the second call to "Schedule()" (makes is a noop), and it switches the arguments to each Exec call according to the transaction progress behind the scenes.
```

Comment:
mult-transaction -> multi-transactio
(makes is a noop) -> (makes it a noop)
In general I do not fully understand this last paragraph.

Reply:
rewritten,